flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/table/KafkaConnectorOptions.java

```diff
@@ -158,6 +158,13 @@
                     .withDescription(
                             "Optional offsets used in case of \"specific-offsets\" startup mode");
 
+    public static final ConfigOption<String> SCAN_BOUNDED_SPECIFIC_OFFSETS =
+            ConfigOptions.key("scan.bounded.specific-offsets")
+                    .stringType()
+                    .noDefaultValue()
+                    .withDescription(
+                            "When all partitions have reached their stop offsets, the source will exit");
+
     public static final ConfigOption<Long> SCAN_STARTUP_TIMESTAMP_MILLIS =
```

Comment:
I think there should be three end modes: latest-offset, timestamp and specific-offsets. Only specific-offsets may not be enough.

Reply:
hi @ruanhang1993  Can you briefly describe the application scenario, if there is a need for this, you can split out several tasks for implementation

Reply:
`latest-offset` may be covered by `timestamp`. So `timestamp` and `specific-offsets` are enough.

IMO `latest-offset` is designed to consume until the latest offsets at the job start time. Maybe this setting already is covered by the `timestamp`.

Reply:
hi @ruanhang1993 You can see if my idea is correct:
First, there are two options for the start read position, an offset and a timestamp.
When the start read position is an offset, it can be either specified or currently consumed, and your end position corresponds to `scan.bounded.stop-offsets`.
When your start bit is a timestamp, your end position should also go to the corresponding timestamp.
So the logic can be changed to: 
with a bounded offset set, first go to determine if `scan.startup.mode` is set, if not go to the logic of `scan.bounded.stop-offsets`, if it is set see if it is a timestamp, if not go to the logic of `scan.bounded. stop-offsets` logic, if it is a timestamp then `scan.bounded.stop-timestamp` logic

Reply:
@ruanhang1993  Do you think the above logic is OK, if so I will make changes

Reply:
`scan.startup.mode` has nothing to do with end logic.

I suggest to provide a `scan.bounded.stop-mode` table option. 
If it is `timestamp`, use `TimestampOffsetsInitializer` for the `stoppingOffsetsInitializer` in the `KafkaSourceBuilder`. 
If it is `specific-offsets`, use `SpecifiedOffsetsInitializer` for the `stoppingOffsetsInitializer` in the `KafkaSourceBuilder`.
Or else it will be the `NoStoppingOffsetsInitializer` by default.

Besides, we need to check the corresponding setting for different modes.

Reply:
I will make changes as soon as possible

Reply:
Perhaps you can set both ` bounedness` and `Bouned `offset like this

```
public KafkaSourceBuilder<OUT> setBounded(OffsetsInitializer stoppingOffsetsInitializer) {
        this.boundedness = Boundedness.BOUNDED;
        this.stoppingOffsetsInitializer = stoppingOffsetsInitializer;
        return this;
    }
```

This allows to set `InstreamingMode `and `InbatchMode`

Reply:
@ruanhang1993 You said `latest-offset may be covered by timestamp ` I don't understand what it means.
What effect does he have on the bounded offse？

Reply:
@nyingping 
![图片](https://user-images.githubusercontent.com/60029759/149496701-c2870be2-1ffa-4082-ae0e-ae7237eb33a7.png)
The call here is `setBounded`, which sets the data source to be bounded directly

Reply:
@ruanhang1993 Can you give me an example?

Reply:
> @ruanhang1993 Can you give me an example?

I means that the user may not need the `latest-offset` end mode. I am not sure about it.
It is of course better to provide this configuration.